blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+        self.transitions = transitions
+
+        if prototype is None:
+            # By default use Linear (with bias) as a default fork.
+            # This overrides the bad default inside Fork which is without bias.
+            # Yes I know use_bias=True us the defaut, but it is a big deal so
+            # I write it down explicitly.
+            prototype = Linear(use_bias=True)
+        depth = len(transitions)
+        self.forks = [Fork(self.normal_inputs(d), prototype=prototype)
+                      for d in range(1, depth)]
+
+        self.children = self.transitions + self.forks
+
+        # Programmatically fix the parameters of the  recurrent decoration
+        setattr(self.apply, 'sequences', transitions[0].apply.sequences)
+        for t in ["states", "contexts", "outputs"]:
+            v = ['%s_%d' % (s, d)
+                 for d, transition in enumerate(transitions)
+                 for s in getattr(transition.apply, t)]
+            setattr(self.apply, t, v)
+
+    def normal_inputs(self, d):
+        transition = self.transitions[d]
+        return [name for name in transition.apply.sequences
+                if 'mask' not in name]
+
+    def _push_allocation_config(self):
+        # Configure the forks that connect the "states" element in the `states`
+        # of one layer to the elements in the `sequences` of the next layer,
+        # excluding "mask".
+        # This involves `get_dim` requests
+        # to the transitions. To make sure that it answers
+        # correctly we should finish its configuration first.
+        for transition in self.transitions:
+            transition.push_allocation_config()
+
+        for d, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[d].get_dim("states")
+            fork.output_dims = self.transitions[d+1].get_dims(
+                fork.output_names)
+
+    @recurrent(sequences=[], states=[], outputs=[], contexts=[])
+    def apply(self, **kwargs):
+        """Apply the stack of transitions.
+
+        Parameters
+        ----------
+        All parameters are of type :class:`~tensor.TensorVariable`.
+
+        Returns
+        -------
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        results = []
+        for d, transition in enumerate(self.transitions):
+            sequences_names = transition.apply.sequences
+            if d == 0:
+                layer_kwargs = dict(
+                    (s, kwargs.get(s)) for s in sequences_names)
+            else:
+                layer_kwargs = dict(zip(
+                    self.normal_inputs(d),
+                    self.forks[d-1].apply(kwargs["states_%d" % (d-1)],
+                                          as_list=True)))
+                if "mask" in sequences_names:
+                    layer_kwargs["mask"] = kwargs.get("mask")
+
+            for t in ["states", "contexts", "outputs"]:
+                suffix = '_%d' % d
```

Comment:
This formatting is not available in Python 3, please use `str.format` or simply `+` operator.